File: python/taco/reader/manifest.py
from __future__ import annotations

from dataclasses import dataclass
from os import fspath
from pathlib import Path
from typing import Any

from . import native
from .source import Location, Source, normalize_sources


@dataclass(frozen=True)
class DatasetResolution:
    sources: tuple[Location, ...]
    collection: dict[str, Any] | None = None
    version: str | None = None
    versions: tuple[str, ...] = ()
    manifest: Location | None = None


def _location(value: str) -> Location:
    return value if "://" in value else Path(value).resolve()


def manifest_candidate(source: Location) -> Location | None:
    candidate = native.manifest_candidate(source)
    if candidate is None:
        return None
    return Path(candidate) if isinstance(source, Path) else candidate


def join_manifest_href(candidate: Location, href: str) -> Location:
    return _location(native.join_manifest_href(fspath(candidate), href))


def resolve_dataset(source: Source) -> DatasetResolution:
    paths = normalize_sources(source)
    if len(paths) != 1:
        return DatasetResolution(paths)

    resolution = native.resolve(paths[0])
    manifest = resolution["manifest"]
    if manifest is None:
        return DatasetResolution(paths)
    return DatasetResolution(
        (_location(resolution["source"]),),
        collection=resolution["collection"],
        version=resolution["version"],
        versions=tuple(resolution["versions"]),
        manifest=manifest if isinstance(paths[0], str) else Path(manifest),
    )


__all__ = [
    "DatasetResolution",
    "join_manifest_href",
    "manifest_candidate",
    "resolve_dataset",
]

File: julia/src/reader/manifest.jl
import JSON3


struct DatasetResolution
    sources::Vector{String}
    collection::Union{Nothing,Dict{String,Any}}
    version::Union{Nothing,String}
    versions::Vector{String}
    manifest::Union{Nothing,String}
    levels::Vector{String}
end


DatasetResolution(sources) = DatasetResolution(sources, nothing, nothing, String[], nothing, String[])


function _plain(value::JSON3.Object)
    result = Dict{String,Any}()
    for (key, item) in pairs(value)
        result[String(key)] = _plain(item)
    end
    return result
end


_plain(value::JSON3.Array) = Any[_plain(item) for item in value]
_plain(value) = value


_is_url(source) = occursin(r"^[A-Za-z][A-Za-z0-9+.-]*://", source)
_local_path(path) = _is_url(path) ? path : normpath(path)


_manifest_candidate(source) = _native_manifest_candidate(source)
_join_manifest_href(candidate, href) = _local_path(_native_join_manifest_href(candidate, href))


function _resolve_dataset(source)
    sources = _normalize_sources(source)
    length(sources) == 1 || return DatasetResolution(sources)

    resolution = _native_resolve(only(sources))
    resolution["manifest"] === nothing && return DatasetResolution(sources)
    collection = resolution["collection"]
    return DatasetResolution(
        [_local_path(String(resolution["source"]))],
        _plain(collection),
        String(resolution["version"]),
        String.(resolution["versions"]),
        String(resolution["manifest"]),
        String.(collect(keys(collection["taco:metadata"]))),
    )
end
